sklearn/utils/_testing.py

```diff
@@ -387,6 +395,63 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
         raise AssertionError("%s not raised by %s" % (names, function.__name__))
 
 
+def assert_allclose(
+    actual, desired, rtol=None, atol=0.0, equal_nan=True, err_msg="", verbose=True
+):
+    """
+    Adaptation of numpy.testing.assert_allclose to have tolerances
+    be set based on the input arrays' dtypes.
```

Comment:
```suggestion
    """dtype-aware variant of numpy.testing.assert_allclose

    This variant introspect the least precise floating point dtype
    in the input argument and automatically sets the relative tolerance
    parameter to 1e-4 float32 and use 1e-7 otherwise (typically float64
    in scikit-learn).

    `atol` is always left to 0.0 by default. It should be adjusted manually
    to an assertion-specific value in case there are null values expected
    in `desired`.

    The aggregate tolerance is ``atol + rtol * abs(desired)``.
```